File: instautils/src/main/java/com/sanjaydevtech/instautils/InstaDownloader.java
package com.sanjaydevtech.instautils;

import android.app.Activity;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.target.CustomTarget;
import com.bumptech.glide.request.transition.Transition;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.IOException;

/**
 * Downloader Class to download insta posts
 *
 * @author Sanjay Developer
 * @version 1.0.1
 */
public class InstaDownloader {
    private Activity activity;
    private InstaResponse response;
    private static final String TAG = InstaDownloader.class.getSimpleName();

    /**
     * Public constructor
     *
     * @param activity Context of the activity
     */
    public InstaDownloader(Activity activity) {
        this.activity = activity;
    }

    /**
     * To attach the InstaResponse Listener
     *
     * @param response Attach the InstaResponse Listener
     */
    public void setResponse(InstaResponse response) {
        this.response = response;
    }

    /**
     * Get the url of the post
     *
     * @param url URL of that post
     * @throws NullPointerException NullPointerException is thrown if InstaResponse is not attached
     */
    public void get(final String url) throws NullPointerException {
        if (response == null) {
            throw new NullPointerException("No InstaResponse Listener is attached");
        }
        new Thread() {
            public void run() {
                try {
                    Document document = Jsoup.connect(url).userAgent("Mozilla/5.0").get();
                    Elements metas = document.getElementsByTag("meta");
                    String img = null, vid = null, type = null;
                    for (Element meta : metas) {
                        String property = meta.attr("property");
                        if (property.equals("og:image")) {
                            img = meta.attr("content");
                            continue;
                        }
                        if (property.equals("og:video")) {
                            vid = meta.attr("content");
                            continue;
                        }
                        if (property.equals("og:type")) {
                            type = meta.attr("content");
                        }
                    }
                    if (type == null || (!type.equals("video") && !type.equals("instapp:photo"))) {
                        activity.runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                response.onError(new NullPointerException("No data resource found"));
                            }
                        });
                    } else {
                        if (type.equals("instapp:photo")) {
                            if (img != null) {
                                final String finalImg = img;
                                activity.runOnUiThread(new Runnable() {
                                    @Override
                                    public void run() {
                                        response.onResponse(new InstaPost(finalImg, InstaPost.INSTA_IMAGE, finalImg));
                                    }
                                });
                            } else {
                                activity.runOnUiThread(new Runnable() {
                                    @Override
                                    public void run() {
                                        response.onError(new NullPointerException("No data resource found"));
                                    }
                                });
                            }
                        } else {
                            if (img != null && vid != null) {
                                final String finalImg1 = img;
                                final String finalVid = vid;
                                activity.runOnUiThread(new Runnable() {
                                    @Override
                                    public void run() {
                                        response.onResponse(new InstaPost(finalVid, InstaPost.INSTA_VIDEO, finalImg1));
                                    }
                                });
                            } else {
                                activity.runOnUiThread(new Runnable() {
                                    @Override
                                    public void run() {
                                        response.onError(new NullPointerException("No data resource found"));
                                    }
                                });
                            }
                        }

                    }

                } catch (final IOException e) {
                    activity.runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            response.onError(e);
                        }
                    });
                }

            }


        }.start();
    }


    /**
     * Retrieve the bitmap of the image post or thumbnail of video post
     *
     * @param post       InstaPost object retrieved from InstaResponse
     * @param instaImage Listener to handle bitmap
     * @throws NullPointerException If InstaImage Listener is not attached
     */
    public void getBitmap(InstaPost post, final InstaImage instaImage) throws NullPointerException {
        if (instaImage == null) {
            throw new NullPointerException("No InstaImage listener attached");
        }
        String imgUrl = post.getThumbnailUrl();
        Glide.with(activity)
                .asBitmap()
                .load(imgUrl)
                .into(new CustomTarget<Bitmap>() {
                    @Override
                    public void onResourceReady(@NonNull Bitmap resource, @Nullable Transition<? super Bitmap> transition) {
                        instaImage.onBitmapLoaded(resource);
                    }

                    @Override
                    public void onLoadCleared(@Nullable Drawable placeholder) {
                    }
                });
    }

    /**
     * Post to ImageView
     *
     * @param post      InstaPost object
     * @param imageView To which view that image has to set
     */
    public void setImage(InstaPost post, ImageView imageView) {
        String imgUrl = post.getThumbnailUrl();
        Glide.with(activity)
                .load(imgUrl)
                .into(imageView);
    }
}

File: instautils/src/main/java/com/sanjaydevtech/instautils/InstaDownloader.kt
package com.sanjaydevtech.instautils

import android.content.Context
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.util.Log
import android.widget.ImageView
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.lifecycleScope
import com.bumptech.glide.Glide
import com.bumptech.glide.request.target.CustomTarget
import com.bumptech.glide.request.transition.Transition
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.jsoup.Jsoup
import java.io.IOException
import java.util.regex.Pattern

/**
 * Downloader Class to download insta posts
 *
 * @author Sanjay Developer
 * @version 1.2.0
 */
class InstaDownloader internal constructor(private val context: Context?, private val scope: CoroutineScope, private val response: InstaResponse) {

    constructor(activity: FragmentActivity, response: InstaResponse) : this(activity, activity.lifecycleScope, response)
    constructor(fragment: Fragment, response: InstaResponse) : this(fragment.context, fragment.viewLifecycleOwner.lifecycleScope, response)

    /**
     * Get the url of the post
     *
     * @param url URL of that post
     */
    fun get(url: String) {
        scope.launch(Dispatchers.IO) {
            try {
                val document = Jsoup.connect(url).userAgent("Mozilla/5.0").get()
                val metas = document.getElementsByTag("meta")
                var img: String? = null
                var vid: String? = null
                var type: String? = null
                for (meta in metas) {
                    val property = meta.attr("property")
                    if (property == "og:image") {
                        img = meta.attr("content")
                        continue
                    }
                    if (property == "og:video") {
                        vid = meta.attr("content")
                        continue
                    }
                    if (property == "og:type") {
                        type = meta.attr("content")
                    }
                }
                if (type == null || type != "video" && type != "instapp:photo") {
                    withContext(Dispatchers.Main) { response.onResponse(InstaTask(null, IllegalArgumentException("No data resource found"))) }
                } else {
                    if (type == "instapp:photo") {
                        if (img != null) {
                            withContext(Dispatchers.Main) { response.onResponse(InstaTask(InstaPost(img, InstaPost.INSTA_IMAGE, img), null)) }
                        } else {
                            withContext(Dispatchers.Main) { response.onResponse(InstaTask(null, IllegalArgumentException("No data resource found"))) }
                        }
                    } else {
                        if (img != null && vid != null) {
                            withContext(Dispatchers.Main) { response.onResponse(InstaTask(InstaPost(vid, InstaPost.INSTA_VIDEO, img), null)) }
                        } else {
                            withContext(Dispatchers.Main) { response.onResponse(InstaTask(null, IllegalArgumentException("No data resource found"))) }
                        }
                    }
                }
            } catch (e: IOException) {
                e.printStackTrace()
                withContext(Dispatchers.Main) { response.onResponse(InstaTask(null, e)) }
            }
        }
    }

    /**
     * Retrieve the bitmap of the image post or thumbnail of video post
     *
     * @param post       InstaPost object retrieved from InstaResponse
     * @param instaImage Listener to handle bitmap
     */
    fun getBitmap(post: InstaPost, instaImage: InstaImage) {
        val imgUrl = post.thumbnailUrl
        context?.let {
            Glide.with(it)
                    .asBitmap()
                    .load(imgUrl)
                    .into(object : CustomTarget<Bitmap?>() {
                        override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap?>?) {
                            instaImage.onBitmapLoaded(resource)
                        }

                        override fun onLoadCleared(placeholder: Drawable?) {}
                    })
        } ?: run {
            Log.e(TAG, "method: getBitmap can't set image.. [REASON] Context is null")
        }
    }

    /**
     * Post to ImageView
     *
     * @param post      InstaPost object
     * @param imageView To which view that image has to set
     */
    fun setImage(post: InstaPost, imageView: ImageView) {
        val imgUrl = post.thumbnailUrl
        context?.let {
            Glide.with(it)
                    .load(imgUrl)
                    .into(imageView)
        } ?: run {
            Log.e(TAG, "method: setImage can't set image.. [REASON] Context is null")
        }
    }

    /**
     * To retrieve Instagram profile
     *
     * @param url      Url of the instagram profile
     */
    fun getDP(url: String) {
        scope.launch(Dispatchers.IO) {
            try {
                val document = Jsoup.connect(url).userAgent("Mozilla/5.0").get()
                val scripts = document.getElementsByTag("script")
                for (script in scripts) {
                    for (dataNode in script.dataNodes()) {
                        val scriptData = dataNode.wholeData.trim { it <= ' ' }
                        if (scriptData.startsWith("window._sharedData")) {
                            var profilePicHD = matchPattern(scriptData, PROFILE_HD_PATTERN)
                            var profilePic = matchPattern(scriptData, PROFILE_PATTERN)
                            if (profilePicHD != null) {
                                profilePicHD = profilePicHD.replace(NOISE, "&")
                            }
                            if (profilePic != null) {
                                profilePic = profilePic.replace(NOISE, "&")
                            }
                            withContext(Dispatchers.Main) { response.onResponse(InstaTask(InstaPost(profilePicHD!!, InstaPost.INSTA_PROFILE, profilePic!!), null)) }
                            return@launch
                        }
                    }
                }
                withContext(Dispatchers.Main) { response.onResponse(InstaTask(null, IllegalArgumentException("No data resource found"))) }

            } catch (e: IOException) {
                withContext(Dispatchers.Main) { response.onResponse(InstaTask(null, e)) }
            }
        }
    }

    private fun matchPattern(data: String, patTxt: String): String? {
        val pattern = Pattern.compile(patTxt)
        val matcher = pattern.matcher(data)
        val patMatch = matcher.find()
        return if (!patMatch) {
            null
        } else matcher.group(1)
    }

    companion object {
        private val TAG = InstaDownloader::class.simpleName
        private const val DP_URL = "^(https://(www\\.)?instagram\\.com/[0-9a-zA-Z_.]+)"
        private const val NOISE = "\\u0026"
        private const val PROFILE_HD_PATTERN = "\"profile_pic_url_hd\":\"([^\"]*)\""
        private const val PROFILE_PATTERN = "\"profile_pic_url\":\"([^\"]*)\""
    }
}